### PR TITLE
Sql Editor

### PR DESCRIPTION
**Pull request by mohdali**
Adds support for SQL syntax which can be used in executing SQL related activities.

- [X] Provides code highlighting using Monaco editor SQL syntax highlighting.

- [X] Provides code completion for simple workflow variables (string and numeric types) using the syntax of `@variable`

- [x] Provides Expression handler to replace `@variable` with the string or numeric value before executing the activity.

This PR also ensures Monaco editor is initialized only once to avoid duplicate completion handlers by using a Mutex promise.

image

### CODE REVIEW

File: src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco-utils.ts [hunk 1]
```diff
@@ -8,30 +9,106 @@ export interface Monaco {
...
+
+let isInitialized: boolean;
+const mutex = new Mutex();
+
+
+
+export async function initializeMonacoWorker(libPath?: string): Promise<Monaco> {
 
-export function initializeMonacoWorker(libPath?: string): Promise<Monaco> {
+  return await mutex.runExclusive(async () => {
```
Comment (sfmskywalker) [inline src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco-utils.ts:26]: Awesome!

File: src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco-utils.ts [hunk 2]
```diff
@@ -8,30 +9,106 @@ export interface Monaco {
...
+      for (let i = 0; i < keywords.length; i++) {
+        autocompleteProviderItems.push({ 'label': keywords[i], kind: monaco.languages.CompletionItemKind.Keyword });
+      }
+
+      return { suggestions: autocompleteProviderItems };
+    }
+  });
+}
+
+function registerSQL(monaco: any) {
```
Comment (sfmskywalker) [inline src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco-utils.ts:92]: The term `SQL` should be renamed to `Sql` (in code only - not in comments etc.). The reason is that according to the naming convention we use, only two-letter acronyms (e.g. `IO`, `IP`) should be all-caps. Longer acronyms such as `Sql`, `Ado`, `Odbc` etc. should be pascal/camel cased (depending on the rules for classes, fields, locals etc.).

**Review by sfmskywalker — CHANGES_REQUESTED** [review]
This looks great! The main thing remaining is to move the SQL expression handler to its own project.

File: src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco-utils.ts (hunk 2 above)
Comment (mohdali) [reply src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco-utils.ts]: Updated to `registerSql`, also, changed `editorVariables` to `EditorVariables`

File: src/core/Elsa.Core/Expressions/SQLHandler.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Elsa.Services.Models;
+
+namespace Elsa.Expressions
+{
+    public class SQLHandler : IExpressionHandler
```
[the remark this answers is not in the document]
Comment (mohdali) [reply src/core/Elsa.Core/Expressions/SQLHandler.cs]: Done. Please check it.

File: src/core/Elsa.Core/Expressions/SyntaxNames.cs [hunk 4]
```diff
@@ -8,5 +8,6 @@ public static class SyntaxNames
         public const string Json = "Json";
         public const string Variable = "Variable";
         public const string Output = "Output";
+        public const string SQL = "SQL";
```
Comment (sfmskywalker) [inline src/core/Elsa.Core/Expressions/SyntaxNames.cs:11]: Should be `public const string Sql = "SQL";`

File: src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco-utils.ts [hunk 5]
```diff
@@ -8,30 +9,106 @@ export interface Monaco {
...
+  type: string;
+}
+
+export var EditorVariables: Array<EditorVariable> = [];
+
+let isInitialized: boolean;
+const mutex = new Mutex();
+
+
+
```
Comment (sfmskywalker) [inline src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco-utils.ts:23]: Looks like some extraneous blanks got in.

File: src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco.tsx [hunk 6]
```diff
@@ -15,17 +15,18 @@ export interface MonacoValueChangedArgs {
...
-  @Prop({attribute: 'single-line', reflect: true}) singleLineMode: boolean = false;
+  @Prop({ attribute: 'single-line', reflect: true }) singleLineMode: boolean = false;
   @Prop() padding: string;
-  @Event({eventName: 'valueChanged'}) valueChanged: EventEmitter<MonacoValueChangedArgs>;
+  @Event({ eventName: 'valueChanged' }) valueChanged: EventEmitter<MonacoValueChangedArgs>;
 
   container: HTMLElement;
   editor: any;
 
+
```
Comment (sfmskywalker) [inline src/designer/elsa-workflows-studio/src/components/controls/elsa-monaco/elsa-monaco.tsx:29]: Remove blank.

File: src/scripting/Elsa.Scripting.Sql/Expression/SQLHandler.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
+using System;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Elsa.Expressions;
+using Elsa.Services.Models;
+
+namespace Elsa.Scripting.Sql.Expressions
+{
+    public class SQLHandler : IExpressionHandler
```
Comment (sfmskywalker) [inline src/scripting/Elsa.Scripting.Sql/Expression/SQLHandler.cs:10]: Should be renamed to `SqlHandler`.

**Review by sfmskywalker — CHANGES_REQUESTED** [review]
I only noticed some small naming issues, nothing major. Once these are updated we can merge this beauty in 😃

**Review by sfmskywalker — APPROVED** [review]
Fantastic. Thanks @mohdali !

**Pull request merged**